### ISSUE TITLE
Incorrect detection of checksum failures for downloaded pieces in 4.0.0-beta.1

### ISSUE THREAD

**Issue by hauuau** [transmission/transmission#4621]
### What is the issue?

I'm using main branch of transmission-daemon and I have noticed that torrents which are marked completed by Transmission often enough have some occasional corruption. After forced verification Transmission notices corruption and redownloads missing pieces.
I've also noticed that those torrents which exhibit corruption have reports about failed checksum checks of pieces during their original download.

I'm not versed enough with Transmission architecture to make a patch but it seems some logic of checksum check is flawed:
https://github.com/transmission/transmission/blob/770da79cf39e41772bc46dfe61f671cd7a1f5de8/libtransmission/torrent.cc#L2391-L2402.
It seems to me that Transmission probably should mark piece for redownloading using something like removePiece after failing checksum check but there is nothing like that there.

This bug may also be related to #3429, #3107, #820, #4178 or similar bugs.

### Which application of Transmission?

transmission-daemon

### Which version of Transmission?

main

**Comment by Deathcow** [comment]
I've noticed this bug too, but it happens very rarely for me and is difficult to catch or replicate, but behaviour is exactly like this: torrent download finishes 100%, but verify shows corrupted pieces, which are then re-downloaded. At some point I was wondering if my storage is to blame, but looks like it might be a transmission bug.

**Comment by ckerr** [comment]
> It seems to me that Transmission probably should mark piece for redownloading using something like removePiece after failing checksum check but there is nothing like that there.

At this point there's no piece to remove; it doesn't get saved until it's passed its checksum test.

The rest of this issue doesn't seem different from #3429, so I'm closing this as a duplicate.

**Comment by ckerr** [comment]
Hmm is that correct? I’ll check again.

**Comment by hauuau** [comment]
> it doesn't get saved until it's passed its checksum test.

How does it actually happen? It seems to me that if that was the case then there should be some code which does that in case of checkPiece returning true in that block. But there is only a call to tr_torrentPieceCompleted which doesn't do much besides renaming a file if every block of a file is marked in some sort of a bitmap.

And that checksum check also happens inside a code block conditioned by tor->hasPiece. Shouldn't that condition be only true if Transmission thinks that it actually has that piece at that point?

**Comment by killemov** [comment]
@ckerr wrote:
> It doesn't get saved until it's passed its checksum test.

And then it still might get corrupted during write (mostly disk full) or on disk.

**Comment by Coeur** [comment]
There is a "**Help Wanted**" sign on #4178, so if you want to help, all you have to do is to add a settings to make my code change optional and then you won't have to worry about torrent corruption anymore. :)

**Comment by hauuau** [comment]
> There is a "**Help Wanted**" sign on #4178 

I think that #4178 is just a workaround that will mask underlying issue. Fixing it properly seems like much better idea.

I have also managed to reliably reproduce this issue using web seeding. The behavior is the same between webseeds and regular ones but it's quite easy to introduce deliberate checksum failures with webseeds.
This script should work as reproducer with transmission-daemon running in the background:

```bash
#!/bin/bash

npm install -g http-server

SEED_PORT=$(shuf -i 10000-65000 -n1)

while true ; do printf "\xAA"; done | dd conv=notrunc of=test.file bs=1c count=64K
transmission-create -o test.file.torrent -s 16 --private -t http://127.0.0.1:${SEED_PORT}/ --anonymize --webseed http://127.0.0.1:${SEED_PORT}/test.file test.file
while true ; do printf "\xBB"; done | dd conv=notrunc of=test.file bs=1c count=48K

http-server -p ${SEED_PORT} &
HTTP_SERVER_PID=$!

transmission-remote -a test.file.torrent
sleep 15
transmission-remote -F n:test.file -i -ic

kill ${HTTP_SERVER_PID}
```

Because first 48K of the `test.file` are corrupted I expect Transmission to notice it and output that it doesn't have a full file and is missing first three pieces.
But instead it outputs the following (without irrelevant info):
```
NAME
  Name: test.file
  Hash: 3da0d84d2d4e1696d6a8a198a3ac3fdb460f804e

TRANSFER
  State: Idle
  Percent Done: 100%
  Have: 65.54 kB (65.54 kB verified)
  Availability: 100%
  Total size: 65.54 kB (65.54 kB wanted)

ORIGINS
  Public torrent: No
  Piece Count: 4
  Piece Size: 16.00 KiB

  1111

```

Transmission properly logs that it had checksum failures and they are recorded in the `corruptEver` field of the torrent info while doing this but the resulting file is corrupted and contains `0xBB` in the first 48K instead of `0xAA`. And there is just no way for Transmission to obtain non-corrupted file at that point anyway.

**Comment by ckerr** [comment]
That is fantastic info!

**Comment by ckerr** [comment]
https://github.com/transmission/transmission/pull/4665 should help with this issue. It's a very simple change and I think it will work, but won't be able to test until later.

If someone else wants to give it a try please let me know how it goes

**Comment by stevenengler** [comment]
Since it seems like the seeder can control the exact contents of the file that transmission downloads, and transmission will say the download completed even when the checksum verifications fail, this seems like a security issue worthy of a CVE and possibly backports. Does this bug exist in the 2.x or 3.x versions?

**Comment by ckerr** [comment]
> Is this bug reproducible in the 2.x or 3.x versions?

https://github.com/transmission/transmission/blob/bb6b5a062ee594dfd4b7a12a6b6e860c43849bfd/libtransmission/torrent.c#L2731 from 3.0.0 indicates this was introduced during the C++ refactoring that led up to 4.0.0-beta.1, but your could confirm with hauuau's test case.

**Comment by diligiant** [comment]
> I think that #4178 is just a workaround that will mask underlying issue. Fixing it properly seems like much better idea.
It's a gonna come handy checking #4665 as if I don't see "incomplete completed" torrents for a few days, I will consider @ckerr commit a winning strike.

**Comment by ckerr** [comment]
This seems to be working for me, but I want to make sure this issue is taken seriously. If anyone else is still seeing this please lmk.

**Comment by diligiant** [comment]
It went from most torrents with large files having a problem to none since applying the patch. Couldn’t be happier. thank you.

**Comment by ckerr** [comment]
Fixed by #4665
